src/Functions/array/mapPopulateSeries.cpp

```diff
@@ -32,85 +34,204 @@ class FunctionMapPopulateSeries : public IFunction
+        if (max_key_type)
         {
-            DataTypePtr max_key_type = arguments[2];
             WhichDataType which_max_key(max_key_type);
 
             if (which_max_key.isNullable())
                 throw Exception(
                     "Max key argument in arguments of function " + getName() + " can not be Nullable",
                     ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
 
-            if (keys_type->getTypeId() != max_key_type->getTypeId())
+            if (key_type->getTypeId() != max_key_type->getTypeId())
                 throw Exception("Max key type in " + getName() + " should be same as keys type", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
         }
+    }
+
+    DataTypePtr getReturnTypeForTuple(const DataTypes & arguments) const
+    {
+        if (arguments.size() < 2)
+            throw Exception(getName() + " accepts at least two arrays for key and value", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (arguments.size() > 3)
+            throw Exception("too many arguments in " + getName() + " call", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        const DataTypeArray * key_array_type = checkAndGetDataType<DataTypeArray>(arguments[0].get());
+        const DataTypeArray * val_array_type = checkAndGetDataType<DataTypeArray>(arguments[1].get());
+
+        if (!key_array_type || !val_array_type)
+            throw Exception(getName() + " accepts two arrays for key and value", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        const auto & key_type = key_array_type->getNestedType();
+
+        if (arguments.size() == 3)
+            this->checkTypes(key_type, arguments[2]);
+        else
+            this->checkTypes(key_type, nullptr);
 
         return std::make_shared<DataTypeTuple>(DataTypes{arguments[0], arguments[1]});
     }
 
+    DataTypePtr getReturnTypeForMap(const DataTypes & arguments) const
+    {
+        const auto * map = assert_cast<const DataTypeMap *>(arguments[0].get());
+        if (arguments.size() == 1)
+            this->checkTypes(map->getKeyType(), nullptr);
+        else if (arguments.size() == 2)
+            this->checkTypes(map->getKeyType(), arguments[1]);
+        else
+            throw Exception("too many arguments in " + getName() + " call", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        return std::make_shared<DataTypeMap>(map->getKeyType(), map->getValueType());
+    }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.empty())
+            throw Exception(getName() + " accepts at least one map", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        if (arguments[0]->getTypeId() == TypeIndex::Array)
+            return getReturnTypeForTuple(arguments);
+        else if (arguments[0]->getTypeId() == TypeIndex::Map)
+            return getReturnTypeForMap(arguments);
+        else
+            throw Exception(getName() + " only accepts maps", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
```

Comment:
It accepts both maps and arrays, so maybe we have to reword. Btw we now have a more convenient syntax for exceptions:
`Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "The function {} accepts only arrays and maps, but '{}' is given", getName(), arguments[0]-><type name or something>"`.
In this case, the message is formatted using the `fmt` library.

Reply:
Maybe it's even better to format all the arguments into the error message, because guessing what's wrong can be a pain. ClickHouse is bad at this...

Reply:
Changed this message and others to this type of creating Exception.